kvpress/presses/kvzip_press.py

```diff
@@ -0,0 +1,358 @@
+# SPDX-FileCopyrightText: Copyright (c) 1993-2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+import logging
+import math
+from contextlib import contextmanager
+from dataclasses import dataclass
+from types import MethodType
+from typing import Generator
+
+import torch
+from torch import nn
+from transformers import AutoTokenizer, Gemma3ForCausalLM, PreTrainedModel, PreTrainedTokenizer, QuantizedCache
+from transformers.models.gemma3.modeling_gemma3 import Gemma3Attention
+from transformers.models.llama.modeling_llama import rotate_half
+from transformers.models.phi3.modeling_phi3 import Phi3Attention
+from transformers.models.qwen3.modeling_qwen3 import Qwen3Attention
+
+from kvpress.presses.base_press import SUPPORTED_MODELS, BasePress
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class KVzipPress(BasePress):
+    """
+    KVzip identifies the importance of KV pairs through context reconstruction,
+    enabling effective query-agnostic KV cache compression.
+
+    In this code, we implement KVzip with minimal changes to this repository.
+    For a fully optimized implementation with actual compression,
+    please refer to the original repository,
+    which also provides a version without runtime compression overhead (at the cost of performance).
+    Original repo (https://github.com/snu-mllab/KVzip).
+
+    Based on KVzip (https://arxiv.org/abs/2505.23416).
+
+    Parameters
+    ----------
+    compression_ratio : float, default=0.0
+        Fraction of key-value pairs to remove during compression.
+    layerwise : bool, default=False
+        Whether to enable uniform compression ratios across layers.
+        When False, while the overall KV cache compression ratio is maintained,
+        each layer has a different compression ratio.
+    n_sink : int, default=4
+        Number of initial tokens to preserve as attention sinks.
+    """
+
+    compression_ratio: float = 0.0
+    layerwise: bool = False
+    n_sink: int = 4
+
+    # The following variables are automatically set in pipeline.py
+    context_length: int = 0
```

Comment:
You can add __post_init__ and set them there - they are not part of the press' init parameters.

Please also initialize all parameters e.g. `self._context_ids` into the post_init method.

Also, please add `logger.warning` in `__post_init__`, informing the user that the press is slower than other presses as there are multiple forward passes.

Reply:
E.g.:
```
            logger.warning(
                "KVzipPress requires multiple forward passes for context reconstruction, "
                "which significantly increases computation time compared to other compression methods. "
                "This is inherent to the KVzip algorithm design."
            )
```